File: Java/1197.java
__________________________________________________________________________________________________
class Solution {
	    public int minKnightMoves(int x, int y) {
	        return (int)knightDistance(x, y);
	    }
		public long knightDistance(long r, long c)
		{
			r = Math.abs(r); c = Math.abs(c);
			if(r + c == 0)return 0;
			if(r + c == 1)return 3;
			if(r == 2 && c == 2)return 4;
			
			long step = Math.max((r+1)/2, (c+1)/2);
			step = Math.max(step, (r+c+2)/3);
			step += (step^r^c)&1;
			return step;
		}
	}

__________________________________________________________________________________________________

__________________________________________________________________________________________________

File: C++/1197.cpp
__________________________________________________________________________________________________
typedef long long ll;
typedef vector<int> VI;
typedef pair<int,int> PII;

#define REP(i,s,t) for(int i=(s);i<(t);i++)
#define FILL(x,v) memset(x,v,sizeof(x))

const int INF = (int)1E9;
#define MAXN 310

int dir[8][2] = {-1,2,1,2,2,1,2,-1,-1,-2,1,-2,-2,1,-2,-1};
int dst[MAXN][MAXN];
class Solution {
public:
    int minKnightMoves(int X, int Y) {
      X = abs(X); Y = abs(Y);
      X += 3; Y += 3;
      queue<PII> q;
      q.push(PII(3, 3));
      REP(i,0,MAXN)REP(j,0,MAXN) dst[i][j] = INF;
      dst[3][3] = 0;
      while (q.size()) {
        PII now = q.front(); q.pop();
        int x = now.first, y = now.second;
        if (x == X && y == Y) return dst[x][y];
        REP(t,0,8) {
          int nx = x + dir[t][0], ny = y + dir[t][1];
          if (nx < 0 || ny < 0 || nx >= MAXN || ny >= MAXN) continue;
          if (dst[nx][ny] == INF) {
            dst[nx][ny] = dst[x][y] + 1;
            q.push(PII(nx, ny));
          }
        }
      }
      return -1;  
    }
};
__________________________________________________________________________________________________

__________________________________________________________________________________________________
